### PR TITLE
ACS-306: ACS Repository: V1 Rest API for content direct access urls

### PR DESCRIPTION
**Pull request by cristianturlica**
- defined endpoints

### CODE REVIEW

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 1]
```diff
@@ -1146,6 +1154,46 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  '/nodes/{nodeId}/request-content-url':
+    post:
+      x-alfresco-since: "7.0"
```
Comment (montgolfiere) [inline src/main/webapp/definitions/alfresco-core.yaml:1159]: Should this be 7.0 or 7.0.0 ?

If 7.0 then we might need to go back and fix others (Version Renditions API) for consistency ?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 2]
```diff
@@ -1146,6 +1154,46 @@ paths:
...
+            Invalid parameter: **nodeId** is not a valid format or **expiryTime** is invalid
+        '401':
+          description: Authentication failed
+        '403':
+          description: Current user does not have permission for **nodeId**
+        '404':
+          description: |
+            **nodeId** does not exist
+        '501':
+          description: The actual ContentStore implementation can't fulfil this request.
```
Comment (ancutam) [inline src/main/webapp/definitions/alfresco-core.yaml:1192]: For consistency, the descriptions for the responses don't usually have a `.` at end

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 1 above)
Comment (cristianturlica) [reply src/main/webapp/definitions/alfresco-core.yaml]: I updated to 7.0.0, thanks for spotting that.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 3]
```diff
@@ -1146,6 +1154,46 @@ paths:
...
+      x-alfresco-since: "7.0"
+      tags:
+        - nodes
+      summary: Generate a direct access content url
+      description: |
+        **Note:** this endpoint is available in Alfresco 7.0 and newer versions.
+
+        Generate a direct access content url for the given **nodeId** with the provided **expiryTime**.
+
+        **Note:** If an **expiryTime** isn't provided then a default of 28800 seconds is used.
```
Comment (ancutam) [inline src/main/webapp/definitions/alfresco-core.yaml:1168]: Maybe add in parenthesis that 28800 seconds mean 8 hours.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 4]
```diff
@@ -1146,6 +1154,46 @@ paths:
...
+      produces:
+        - application/json
+      responses:
+        '200':
+          description: Successful response
+          schema:
+            $ref: '#/definitions/NodeDirectAccessEntry'
+        '400':
+          description: |
+            Invalid parameter: **nodeId** is not a valid format or **expiryTime** is invalid
```
Comment (ancutam) [inline src/main/webapp/definitions/alfresco-core.yaml:1183]: Same as for the others endpoints, invalid if "is not a file", right?

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 5]
```diff
@@ -173,6 +173,14 @@ parameters:
     items:
       type: string
     collectionFormat: csv
+  nodeDirectAccessExpiryTimeParam:
+    name: expiryTime
+    in: query
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:178]: If we're providing parameters for a POST endpoint it should be in the body of the POST, not a query string parameter.

Comment (cristianturlica) [reply src/main/webapp/definitions/alfresco-core.yaml]: Change done

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 6]
```diff
@@ -1146,6 +1146,60 @@ paths:
...
+        - nodes
+      summary: Generate a direct access content url
+      description: |
+        **Note:** this endpoint is available in Alfresco 7.0 and newer versions.
+
+        Generate a direct access content url for the given **nodeId**.  Also, an optional expiry date could be set, so the direct access link would become invalid when the expiry date is reached. For example:
+
+        ```JSON
+         {
+
```
Comment (ancutam) [inline src/main/webapp/definitions/alfresco-core.yaml:1162]: extra line?

Comment (cristianturlica) [reply src/main/webapp/definitions/alfresco-core.yaml]: Removed, thanks.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 7]
```diff
@@ -1146,6 +1146,59 @@ paths:
...
+        **Note:** If an expiry date isn't provided then a default of 28800 seconds (8 hours) link validity time is used if not configured otherwise.
+        **Note:** It is up to the actual ContentStore implementation if it can fulfil this request or not.
+      operationId: request-content-url
+      parameters:
+        - $ref: '#/parameters/nodeIdParam'
+        - in: body
+          name: requestContentUrlBodyCreate
+          description: |
+            Optionally, specify the length of time in seconds that the link is valid for.
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:1174]: We still have the old description in here, same for the other end points too.

**Comment by cristianturlica** [comment]
> 1\. As a developer, I would prefer a lifetime instead of an end date. When I am getting a download link in a web app, I want to set a short timeout like 60sec or alike. And I don't want to deal with time calculations. Also I don't want to deal with TZ problems. But if I have to give the lifetime in terms of a fixed endpoint, then I need to do all the calculations in my frontend web app. And it isn't even guaranteed that I have proper time information there to do this calculation

In this case I will go back to the initial implementation using long for lifetime all the way back to ContentStore.

> 2\. A default lifetime of 8hrs seems pretty long. I would propose something like 5min as a default. And we should mention that the admin can overwrite the default

I will update it to 5 minutes.

> 3\. What if the content store does not support these URLs? Do we still provide the URL in the front end but provide the service ourselves?

We throw an exception (501) when it isn't supported.

I thing we should also return back to the user the lifetimeInSec value as this will allow the user to know it's value when the default is used.

@skopf would you agree to the above?

File: src/main/webapp/definitions/alfresco-core.yaml (hunk 7 above)
Comment (cristianturlica) [reply src/main/webapp/definitions/alfresco-core.yaml]: Updated now

**Comment by cristianturlica** [comment]
Changes after feedback: expiresAt (date) and validFor (long time in seconds) can now be set

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 8]
```diff
@@ -1146,6 +1146,67 @@ paths:
...
+    post:
+      x-alfresco-since: "7.0.0"
+      tags:
+        - nodes
+      summary: Generate a direct access content url
+      description: |
+        **Note:** this endpoint is available in Alfresco 7.0 and newer versions.
+
+        Generate a direct access content url for the given **nodeId**.  
+        Optionaly the expiry at date could be set, so the direct access link would become invalid when the expiry date is reached. For example:
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:1159]: typo "optionaly" -> "optionally"

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 9]
```diff
@@ -1146,6 +1146,67 @@ paths:
...
+
+        ```JSON
+         {
+           "expiresAt": "2017-03-23T23:00:00.000+0000"
+         }
+        ```
+        Or optionaly the validFor (in seconds) could be set, so the direct access link is valid for that length of time.
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:1166]: typo "optionaly" -> "optionally" again, I won't add anymore comments but there are others in the file.

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 10]
```diff
@@ -8759,6 +9007,14 @@ definitions:
...
           message:
             type: string
+  DirectAccessUrlBodyCreate:
+  type: object
+  properties:
+    expiresAt:
+      type: string
+      format: date-time
+    validFor:
+      type: string
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:9017]: validFor should have a type of integer and set a minimum of 1 (see maxItemsParam for an example)

File: src/main/webapp/definitions/alfresco-core.yaml [hunk 11]
```diff
@@ -9235,66 +9491,21 @@ definitions:
...
-        description: the human-readable constraint description
+        description: The direct access URL of a binary content
+      expiresAt:
         type: string
-      parameters:
-        type: object
-        additionalProperties:
-          type: object
\ No newline at end of file
+        format: date-time
```
Comment (gavincornwell) [inline src/main/webapp/definitions/alfresco-core.yaml:9511]: It's probably worth also adding a description for the expiresAt property

**Pull request merged**